AirCasting/Utils/Microphone/AVMicrophone.swift

```diff
@@ -0,0 +1,96 @@
+// Created by Lunar on 17/05/2022.
+//
+
+import Foundation
+import AVFoundation
+import UIKit
+
+final class AVMicrophone: Microphone {
+    enum AVMicrophoneError: Error {
+        case permissionNotGranted, couldntStartRecorder
+    }
+    private(set) var state: MicrophoneState = .notRecording
+    
+    static private let recordSettings: [String: Any] = [
+        AVFormatIDKey: NSNumber(value: kAudioFormatAppleLossless),
+        AVSampleRateKey: 44100.0,
+        AVNumberOfChannelsKey: 1,
+        AVEncoderAudioQualityKey: AVAudioQuality.min.rawValue
+    ]
+    private let recorder: AVAudioRecorder
+    private let audioSession = AVAudioSession.sharedInstance()
+    private var interruptionHandler: AVSessionInterruptionHandler!
+    
+    init() throws {
+        self.recorder = try AVAudioRecorder(url: URL(fileURLWithPath: "/dev/null", isDirectory: true), settings: AVMicrophone.recordSettings)
+        self.interruptionHandler = .init(observer: self, audioSession: audioSession)
+    }
+    
+    func getCurrentDecibelLevel() -> Double? {
+        recorder.updateMeters()
+        let power = recorder.averagePower(forChannel: 0)
+        var decibels = Double(power + 90.0)
+        (decibels < 0) ? decibels = 0 : nil
+        return decibels
```

Comment:
Oh, great catch!

I'm thinking what would be correct on a domain level. Can microphone read `no value`? What if we try to get current level when the microphone is stopped? Should we `nil`? Should we throw? Should we crash? Or just pretend nothing's happening and return `0`?

I'll think about it, if you have any thoughts please share!

Reply:
The only thought I have right now is not to use `0` when the mic is stopped. I think it would mislead us, as the 0 is a numeric value that could potentially happen during the standard process of recording (possibly not, but who knows)